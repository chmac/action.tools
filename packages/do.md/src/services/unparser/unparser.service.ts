import { Content, ListItem, PhrasingContent, Root, Paragraph } from 'mdast';
import { KEY_VALUE_SEPARATOR } from '../../constants';
import { Section, Task, TaskData } from '../../types';
import unified from 'unified';
import markdown from 'remark-parse';

const toMdastProcessor = unified().use(markdown, { gfm: true });

export const serializeData = (data: TaskData) => {
  // NOTE: This will have an extra space as the first element
  const children: PhrasingContent[] = Object.entries(data).flatMap(
    ([key, value]) => {
      return [
        { type: 'text', value: ' ' },
        {
          type: 'inlineCode',
          value: `${key}${KEY_VALUE_SEPARATOR}${value}`,
        },
      ];
    }
  );

  // NOTE: This check is necessary because the next `.splice()` will apply
  // whether or not there are any elements in the array. If there is data, we
  // want to return an empty array, and not a break.
  if (children.length === 0) {
    return children;
  }

  // Replace the first space with a break instead
  children.splice(0, 1, { type: 'break' });

  return children;
};

export const recursiveTaskToMdast = ({
  tasks,
  taskId,
}: {
  tasks: Task[];
  taskId: string;
}) => {
  const task = tasks.find(task => task.id === taskId);
  if (typeof task === 'undefined') {
    throw new Error(
      'recursiveTaskToMdast() called with invalid taskId #iuLSAf'
    );
  }
  // NOTE: We only allow 1 child paragraph and 1 child list

  const root = toMdastProcessor.parse(task.contentMarkdown) as Root;
  const paragraph = root.children[0] as Paragraph;

  const dataContent = serializeData(task.data);
  paragraph.children.push(...dataContent);

  const childTasks = tasks.filter(task => task.parentId === taskId);
  const hasChildTasks = childTasks.length > 0;

  const children: Content[] = [paragraph];

  if (hasChildTasks) {
    const childListItems = childTasks.map(task =>
      recursiveTaskToMdast({ tasks, taskId: task.id })
    );
    children.push({
      type: 'list',
      // NOTE: We assume that if 1 child task is sequential, they must all be,
      // because this list must be an ordered (numbered) list.
      ordered: childTasks[0].isSequential,
      spread: false,
      children: childListItems,
    });
  }

  const mdast = {
    type: 'listItem',
    checked: task.isTask ? task.finished : null,
    spread: false,
    children,
  } as ListItem;
  return mdast;
};

export const createMdast = ({
  sections,
  tasks,
}: {
  sections: Section[];
  tasks: Task[];
}): Root => {
  const children: Content[] = [];
  sections.forEach(section => {
    if (typeof section.heading !== 'undefined') {
      children.push(section.heading);
    }
    children.push(...section.contents);

    const sectionTasks = tasks.filter(
      task => task.sectionId === section.id && task.parentId === ''
    );

    if (sectionTasks.length === 0) {
      return;
    }

    const listItems = sectionTasks.map(task =>
      recursiveTaskToMdast({ tasks, taskId: task.id })
    );

    children.push({
      type: 'list',
      spread: false,
      // NOTE: We assume that if 1 child task is sequential, they must all be,
      // because this list must be an ordered (numbered) list.
      ordered: sectionTasks[0].isSequential,
      children: listItems,
    });
  });

  return { type: 'root', children };
};
