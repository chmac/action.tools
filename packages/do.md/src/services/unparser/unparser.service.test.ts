import { LocalDate } from '@js-joda/core';
import { mdastToMarkdown } from '../../__fixtures__/markdown.fixtures';
import { createMdast, serializeData } from './unparser.service';

describe('unparser', () => {
  describe('serializeData()', () => {
    it('Serializes #yy5H4E', () => {
      expect(
        serializeData({ id: 'foo', after: LocalDate.of(2020, 2, 24) })
      ).toMatchSnapshot();
    });

    it('Returns empty array for no data #XF6duP', () => {
      expect(serializeData({})).toEqual([]);
    });
  });

  describe('createMdast()', () => {
    it('Creates an empty root without content #kLjEzn', () => {
      expect(
        createMdast({
          sections: [{ id: 'top', contents: [], depth: 0 }],
          tasks: [],
        })
      ).toMatchSnapshot();
    });

    it('Creates an initial task list #73kagJ', () => {
      expect(
        createMdast({
          sections: [{ id: 'top', contents: [], depth: 0 }],
          tasks: [
            {
              id: 'example1',
              finished: false,
              sectionId: 'top',
              isSequential: false,
              isTask: true,
              parentId: '',
              data: {},
              contentMarkdown: 'A task',
            },
          ],
        })
      ).toMatchSnapshot();
    });

    it('Creates a task list that matches expected markdown with one task #ARaqJE', async () => {
      expect(
        await mdastToMarkdown(
          createMdast({
            sections: [{ id: 'top', contents: [], depth: 0 }],
            tasks: [
              {
                id: 'example1',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A task',
              },
            ],
          })
        )
      ).toEqual(`- [ ] A task\n`);
    });
    it('Creates a task list that matches expected markdown with three tasks #LbIvfd', async () => {
      expect(
        await mdastToMarkdown(
          createMdast({
            sections: [{ id: 'top', contents: [], depth: 0 }],
            tasks: [
              {
                id: 'example1',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A task',
              },
              {
                id: 'example2',
                finished: true,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A completed task',
              },
              {
                id: 'example3',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'Another task',
              },
            ],
          })
        )
      ).toEqual(`- [ ] A task\n- [x] A completed task\n- [ ] Another task\n`);
    });

    it('Creates a task list that matches expected markdown with two sections #3yh9IU', async () => {
      expect(
        await mdastToMarkdown(
          createMdast({
            sections: [
              { id: 'top', contents: [], depth: 0 },
              {
                id: 'h1',
                contents: [
                  {
                    type: 'paragraph',
                    children: [
                      {
                        type: 'text',
                        value: 'A paragraph under the heading.',
                      },
                    ],
                  },
                ],
                depth: 1,
                heading: {
                  type: 'heading',
                  depth: 1,
                  children: [{ type: 'text', value: 'A top heading' }],
                },
              },
            ],
            tasks: [
              {
                id: 'example1',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A task',
              },
              {
                id: 'example2',
                finished: true,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A completed task',
              },
              {
                id: 'example3',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'Another task',
              },
              {
                id: 'example4',
                finished: false,
                sectionId: 'h1',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A sectioned task',
              },
              {
                id: 'example5',
                finished: false,
                sectionId: 'h1',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A second sectioned task',
              },
            ],
          })
        )
      ).toEqual(`- [ ] A task
- [x] A completed task
- [ ] Another task

# A top heading

A paragraph under the heading.

- [ ] A sectioned task
- [ ] A second sectioned task
`);
    });

    it('Creates a task list that matches the expected markdown with a nested list #XLbatC', async () => {
      expect(
        await mdastToMarkdown(
          createMdast({
            sections: [
              { id: 'top', contents: [], depth: 0 },
              {
                id: 'h1',
                contents: [
                  {
                    type: 'paragraph',
                    children: [
                      {
                        type: 'text',
                        value: 'A paragraph under the heading.',
                      },
                    ],
                  },
                ],
                depth: 1,
                heading: {
                  type: 'heading',
                  depth: 1,
                  children: [{ type: 'text', value: 'A top heading' }],
                },
              },
            ],
            tasks: [
              {
                id: 'example1',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A task',
              },
              {
                id: 'example2',
                finished: true,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A completed task',
              },
              {
                id: 'example21j',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: 'example2',
                isTask: true,
                data: {},
                contentMarkdown: 'A child task',
              },
              {
                id: 'example22',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: 'example2',
                isTask: true,
                data: {},
                contentMarkdown: 'Another child task',
              },
              {
                id: 'example3',
                finished: false,
                sectionId: 'top',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'Another top level task',
              },
              {
                id: 'example4',
                finished: false,
                sectionId: 'h1',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A sectioned task',
              },
              {
                id: 'example5',
                finished: false,
                sectionId: 'h1',
                isSequential: false,
                parentId: '',
                isTask: true,
                data: {},
                contentMarkdown: 'A second sectioned task',
              },
            ],
          })
        )
      ).toEqual(`- [ ] A task
- [x] A completed task
  - [ ] A child task
  - [ ] Another child task
- [ ] Another top level task

# A top heading

A paragraph under the heading.

- [ ] A sectioned task
- [ ] A second sectioned task
`);
    });
  });
});
