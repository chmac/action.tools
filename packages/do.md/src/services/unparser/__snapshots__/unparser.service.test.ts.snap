// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unparser createMdast() Creates an empty root without content #kLjEzn 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "type": "list",
    },
  ],
  "type": "root",
}
`;

exports[`unparser createMdast() Creates an initial task list #73kagJ 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "checked": false,
          "children": Array [
            Object {
              "type": "text",
              "value": "A task",
            },
          ],
          "type": "listItem",
        },
      ],
      "type": "list",
    },
  ],
  "type": "root",
}
`;

exports[`unparser serializeData() Serializes #yy5H4E 1`] = `
Array [
  Object {
    "type": "break",
  },
  Object {
    "type": "inlineCode",
    "value": "id:foo",
  },
  Object {
    "type": "text",
    "value": " ",
  },
  Object {
    "type": "inlineCode",
    "value": "after:2020-02-24",
  },
]
`;
